src/base/searchengine.h

```diff
@@ -38,10 +40,133 @@
+{
+public:
+    typedef T ComponentType;
+    typedef Version<T,N> this_type;
+
+    Version()
+    {
+        m_components.fill(T());
+    }
+
+    template <class ...Other>
+    Version(Other... components)
+        : m_components {{components...}}
+    {
+    }
+
+    Version(const QString &version)
+    {
+        QStringList parts = version.split(QLatin1Char('.'));
+        if (parts.size() != N)
+            throw std::logic_error("Incorrect number of version components");
+        bool ok = false;
+        for (std::size_t i = 0; i < N; ++i) {
+            m_components[i] = static_cast<T>(parts[i].toInt(&ok));
+            if (!ok)
+                throw std::runtime_error("Can not parse version components");
+        }
+    }
+
+    Version(const QByteArray &version)
+    {
+        QList<QByteArray> parts = version.split('.');
+        if (parts.size() != N)
+            throw std::logic_error("Incorrect number of version components");
+        bool ok = false;
+        for (std::size_t i = 0; i < N; ++i) {
+            m_components[i] = static_cast<T>(parts[i].toInt(&ok));
+            if (!ok)
+                throw std::runtime_error("Can not parse version components");
+        }
+    }
+
+    ComponentType majorNumber() const {
+        static_assert(N >= 1, "Array size is too small");
+        return m_components[0];
+    }
+
+    ComponentType minorNumber() const {
+        static_assert(N >= 2, "Array size is too small");
+        return m_components[1];
+    }
+
+    ComponentType revisionNumber() const {
+        static_assert(N >= 2, "Array size is too small");
+        return m_components[2];
+    }
+
+    ComponentType patchNumber() const {
+        static_assert(N >= 3, "Array size is too small");
+        return m_components[3];
+    }
+
+    ComponentType operator[](std::size_t i) const
+    {
+        return m_components.at(i);
+    }
+
+    QString toString() const
+    {
+        QString res = QString::number((*this)[0]);
+        for (std::size_t i = 1; i < N; ++i) {
+            res += QLatin1Char('.') + QString::number((*this)[i]);
+        }
+        return res;
+    }
+
+    bool operator == (const this_type &other) const
+    {
+        return m_components == other.m_components;
+    }
+
+    bool operator != (const this_type &other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator < (const this_type &other) const
+    {
+        for (std::size_t i = 0; i < N - 1; ++i) {
+            if (m_components[i] > other.m_components[i])
+                return false;
+        }
+        return m_components.back() < other.m_components.back();
+    }
+
+    bool operator > (const this_type &other) const
+    {
+        for (std::size_t i = 0; i < N - 1; ++i) {
+            if (m_components[i] < other.m_components[i])
+                return false;
+        }
+        return m_components.back() > other.m_components.back();
+    }
+private:
+    using ComponentsArray = std::array<unsigned, N>;
```

Comment:
>AFAIK, tuple is array with items of different type.

This won't matter if agrees on some fixed typedef:
`typedef std::tuple<int, int, int, Int> Version;`

>Why?

An ideal `Version` type has the same mathematical properties as `std::tuple<int, int, int, Int>`.
And mainly because `std::tuple` already handles the relational operators just as we wanted. All there's missing is some helper methods to do type conversions (if doing this way).
The `Version` class could be much more simpler, IMO.

But all of this doesn't really matters now, see: https://github.com/qbittorrent/qBittorrent/pull/5264#discussion_r63661534